File: content/develop/use-cases/cache-aside/dotnet/MockPrimaryStore.cs
namespace CacheAsideDemo;

/// <summary>
/// In-memory stand-in for a slow primary database. Each read sleeps for
/// <c>ReadLatencyMs</c> milliseconds so the difference between a cache hit
/// and a miss is visible in the demo UI.
/// </summary>
public class MockPrimaryStore
{
    public int ReadLatencyMs { get; }

    private long _reads;
    private readonly object _lock = new();
    private readonly Dictionary<string, Dictionary<string, string>> _records;

    public MockPrimaryStore(int readLatencyMs = 150)
    {
        ReadLatencyMs = readLatencyMs;
        _records = new Dictionary<string, Dictionary<string, string>>(StringComparer.Ordinal)
        {
            ["p-001"] = new() { ["id"] = "p-001", ["name"] = "Sourdough Loaf",      ["price_cents"] = "650",  ["stock"] = "42"  },
            ["p-002"] = new() { ["id"] = "p-002", ["name"] = "Espresso Beans 250g", ["price_cents"] = "1495", ["stock"] = "120" },
            ["p-003"] = new() { ["id"] = "p-003", ["name"] = "Olive Oil 500ml",     ["price_cents"] = "1200", ["stock"] = "8"   },
            ["p-004"] = new() { ["id"] = "p-004", ["name"] = "Sea Salt Flakes",     ["price_cents"] = "475",  ["stock"] = "60"  },
        };
    }

    public List<string> ListIds()
    {
        var ids = _records.Keys.ToList();
        ids.Sort(StringComparer.Ordinal);
        return ids;
    }

    /// <summary>Slow read of a record. Returns null if id is unknown.</summary>
    public Dictionary<string, string>? Read(string id)
    {
        Thread.Sleep(ReadLatencyMs);
        Interlocked.Increment(ref _reads);
        lock (_lock)
        {
            return _records.TryGetValue(id, out var record)
                ? new Dictionary<string, string>(record, StringComparer.Ordinal)
                : null;
        }
    }

    public bool UpdateField(string id, string field, string value)
    {
        lock (_lock)
        {
            if (!_records.TryGetValue(id, out var record))
            {
                return false;
            }
            record[field] = value;
            return true;
        }
    }

    public long Reads => Interlocked.Read(ref _reads);

    public void ResetReads() => Interlocked.Exchange(ref _reads, 0);
}

File: content/develop/use-cases/cache-aside/java-lettuce/MockPrimaryStore.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicLong;

/**
 * In-memory stand-in for a slow primary database. Each read sleeps for
 * {@code readLatencyMs} milliseconds so the difference between a cache hit
 * and a miss is visible in the demo UI.
 */
public class MockPrimaryStore {

    private final int readLatencyMs;
    private final AtomicLong reads = new AtomicLong();
    private final Map<String, Map<String, String>> records = new HashMap<>();

    public MockPrimaryStore(int readLatencyMs) {
        this.readLatencyMs = readLatencyMs;
        seed("p-001", "Sourdough Loaf", "650", "42");
        seed("p-002", "Espresso Beans 250g", "1495", "120");
        seed("p-003", "Olive Oil 500ml", "1200", "8");
        seed("p-004", "Sea Salt Flakes", "475", "60");
    }

    public int getReadLatencyMs() {
        return readLatencyMs;
    }

    public List<String> listIds() {
        List<String> ids = new java.util.ArrayList<>(records.keySet());
        java.util.Collections.sort(ids);
        return ids;
    }

    /** Slow read of the record. Returns null if id is unknown. */
    public Map<String, String> read(String id) {
        try {
            Thread.sleep(readLatencyMs);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return null;
        }
        synchronized (records) {
            reads.incrementAndGet();
            Map<String, String> record = records.get(id);
            if (record == null) {
                return null;
            }
            return new LinkedHashMap<>(record);
        }
    }

    public boolean updateField(String id, String field, String value) {
        synchronized (records) {
            Map<String, String> record = records.get(id);
            if (record == null) {
                return false;
            }
            record.put(field, value);
            return true;
        }
    }

    public long getReads() {
        return reads.get();
    }

    public void resetReads() {
        reads.set(0);
    }

    private void seed(String id, String name, String priceCents, String stock) {
        Map<String, String> record = new LinkedHashMap<>();
        record.put("id", id);
        record.put("name", name);
        record.put("price_cents", priceCents);
        record.put("stock", stock);
        records.put(id, record);
    }
}
